server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2773,4 +2773,64 @@ private static void trimUnsafeCommits(EngineConfig engineConfig) throws IOExcept
+        final Weight weight = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
+        for (LeafReaderContext leaf : directoryReader.leaves()) {
+            final LeafReader reader = leaf.reader();
+            final Scorer scorer = weight.scorer(leaf);
+            if (scorer == null) {
+                continue;
+            }
+            final DocIdSetIterator docIdSetIterator = scorer.iterator();
+            final NumericDocValues seqNoDV = Objects.requireNonNull(reader.getNumericDocValues(SeqNoFieldMapper.NAME));
+            final NumericDocValues primaryTermDV = Objects.requireNonNull(reader.getNumericDocValues(SeqNoFieldMapper.PRIMARY_TERM_NAME));
+            final NumericDocValues tombstoneDV = reader.getNumericDocValues(SeqNoFieldMapper.TOMBSTONE_NAME);
+            final NumericDocValues versionDV = reader.getNumericDocValues(VersionFieldMapper.NAME);
+            int docId;
+            while ((docId = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                if (primaryTermDV.advanceExact(docId) == false) {
+                    continue; // skip children docs which do not have primary term
+                }
+                final long primaryTerm = primaryTermDV.longValue();
+                if (seqNoDV.advanceExact(docId) == false) {
+                    throw new IllegalStateException("seq_no not found for doc_id=" + docId);
+                }
+                final long seqNo = seqNoDV.longValue();
+                localCheckpointTracker.markSeqNoAsCompleted(seqNo);
+                final boolean isTombstone = tombstoneDV != null && tombstoneDV.advanceExact(docId);
+                final FieldsVisitor fields = new FieldsVisitor(false, SourceFieldMapper.NAME);
+                reader.document(docId, fields);
+                fields.postProcess(engineConfig.getMapperService());
+                if (fields.uid() == null) {
+                    assert isTombstone; // a noop
+                    continue;
+                }
+                final BytesRef uid = new Term(IdFieldMapper.NAME, Uid.encodeId(fields.uid().id())).bytes();
+                if (versionDV == null || versionDV.advanceExact(docId) == false) {
+                    throw new IllegalStateException("version not found for doc_id=" + docId);
+                }
+                final long version = versionDV.longValue();
+                try (Releasable ignored = versionMap.acquireLock(uid)) {
+                    final VersionValue curr = versionMap.getUnderLock(uid);
+                    if (curr == null || curr.term < primaryTerm || (curr.term == primaryTerm && curr.seqNo <= seqNo)) {
+                        if (isTombstone) {
+                            versionMap.putDeleteUnderLock(uid,
+                                new DeleteVersionValue(version, seqNo, primaryTerm, engineConfig.getThreadPool().relativeTimeInMillis()));
```

Comment:
does using the current time stamp keep this delete around unnecessary long in the version map?

Reply:
I added a comment.